File: backend/routes/userRoutes.py
from fastapi import APIRouter, HTTPException, Depends
from schemas.userSchemas import UserBase, UserUpdate
from services import userServices
from services.authServices import get_current_user

router = APIRouter(prefix="/usuarios", tags=["Usuarios"])

@router.get("/", dependencies=[Depends(get_current_user)])
def listar_usuarios():
    return userServices.get_all_users()

@router.get("/me")
def perfil_me(current_user=Depends(get_current_user)):
    return current_user

@router.get("/{user_id}", dependencies=[Depends(get_current_user)])
def buscar_usuario(user_id: str):
    user = userServices.get_user_by_id(user_id)
    if not user:
        raise HTTPException(status_code=404, detail="Usuário não encontrado")
    return user

@router.post("/")
def criar_usuario(user: UserBase):
    try:
        return userServices.create_user(user)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.put("/{user_id}", dependencies=[Depends(get_current_user)])
def atualizar_usuario(user_id: str, user: UserUpdate):
    updated_user = userServices.update_user(user_id, user)
    if not updated_user:
        raise HTTPException(status_code=404, detail="Usuário não encontrado ou sem alterações")
    return updated_user

@router.delete("/{user_id}", dependencies=[Depends(get_current_user)])
def deletar_usuario(user_id: str):
    success = userServices.delete_user(user_id)
    if not success:
        raise HTTPException(status_code=404, detail="Usuário não encontrado")
    return {"message": "Usuário deletado com sucesso"}

File: backend/routes/userRoutes.js
const express = require('express');
const { getAllUsers, getUserById, createUser } = require('../services/userServices');
const { getCurrentUser } = require('../services/authServices');

const router = express.Router();

// Middleware to check authentication
const authenticate = async (req, res, next) => {
  try {
    req.currentUser = await getCurrentUser(req);
    next();
  } catch (error) {
    res.status(401).json({ error: 'Unauthorized' });
  }
};

// List all users
router.get('/', authenticate, async (req, res) => {
  try {
    const users = await getAllUsers();
    res.json(users);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch users' });
  }
});

// Get current user's profile
router.get('/me', authenticate, (req, res) => {
  res.json(req.currentUser);
});

// Get a user by ID
router.get('/:userId', authenticate, async (req, res) => {
  try {
    const user = await getUserById(req.params.userId);
    if (!user) {
      return res.status(404).json({ error: 'User not found' });
    }
    res.json(user);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch user' });
  }
});

// Create a new user
router.post('/', async (req, res) => {
  try {
    const user = await createUser(req.body);
    res.status(201).json(user);
  } catch (error) {
    res.status(400).json({ error: error.message });
  }
});

module.exports = router;